File: 002_reading_writing_videos/CPP/video_write_to_file/video_write_to_file.cpp
// Include Libraries
#include<opencv2/opencv.hpp>
#include<iostream>

// Namespace to nullify use of cv::function(); syntax
using namespace std;
using namespace cv;

int main()
{
    // initialize a video capture object
	VideoCapture vid_capture("Resources/Cars.mp4");

    // Acquire frame width and height with the help of get() method
    // You can replace 3 and 4 with CAP_PROP_FRAME_WIDTH and CAP_PROP_FRAME_HEIGHT
    // They are just enumerations
	int frame_width = static_cast<int>(vid_capture.get(3));
	int frame_height = static_cast<int>(vid_capture.get(4));

    //define frame_size to be used in VideoWriter() argument
	Size frame_size(frame_width, frame_height);
	int fps = 20;

    // Initialize video writer object
	VideoWriter output("Resources/output.avi", VideoWriter::fourcc('M', 'J', 'P', 'G'), fps, frame_size);
	
    while (vid_capture.isOpened())
    {
        // Initialize frame matrix to store frames
        Mat frame;

        // Initialize a boolean to check if the frames are present or not
        bool isSuccess = vid_capture.read(frame);

        // If frames are not there, close it
        if (isSuccess == false)
        {
            cout << "Video camera is disconnected" << endl;
            break;
        }
        // If frames are present
        if(isSuccess == true)
        {
            //display frames
            output.write(frame);
            // display frames
            imshow("Frame", frame);

            // wait for 20 ms between successive frames and break the loop if key q is pressed
            int key = waitKey(20);
            if (key == 'q')
            {
                cout << "Key q is pressed by the user. Stopping the video" << endl;
                break;
            }
        }
    }
   destroyAllWindows();
   vid_capture.release();
   output.release();
   return 0;
}

File: 002_reading_writing_videos/Python/video_write_to_file.py
# Import packages
import cv2

# Create a video capture object
vid_capture = cv2.VideoCapture('Resources/Cars.mp4')

# Print error message if object is not in open state
if(vid_capture.isOpened() == False):
	print("Error opening video file")

# Get height and width of the frame
#CAP_PROP_FRAME_WIDTH = 3, CAP_PROP_FRAME_HEIGHT = 4
frame_width = int(vid_capture.get(3))
frame_height = int(vid_capture.get(4))
frame_size = (frame_width,frame_height)
fps = 20

# Create a video writer object
output = cv2.VideoWriter('Resources/output_video_from_file.avi', cv2.VideoWriter_fourcc('M','J','P','G'), fps, frame_size)

while(vid_capture.isOpened()):
	# vCapture.read() methods returns a tuple, first element is a bool 
	# and the second is frame
	ret, frame = vid_capture.read()

	if ret == True:
		output.write(frame)

		cv2.imshow("Frame",frame)
		# k == 113 is ASCII code for q key. You can try to replace that 
		# with any key with its corresponding ASCII code, try 27 which is for ESCAPE
		key = cv2.waitKey(20)
		if key == ord('q'):
			break
	else:
		print('Stream disconnected')
		break
# Release the video capture and output objects.
vid_capture.release()
output.release()
cv2.destroyAllWindows()